File: src/main/scala/util/scala/WatermarkStrategies.scala
package util.scala

import org.apache.flink.api.common.eventtime.{SerializableTimestampAssigner, WatermarkStrategy}

import java.time.Duration

object WatermarkStrategies {

  val sensorReadingStrategy: WatermarkStrategy[SensorReading] = WatermarkStrategy
    .forBoundedOutOfOrderness(Duration.ofSeconds(5L))
    .withTimestampAssigner(new SerializableTimestampAssigner[SensorReading] {
      override def extractTimestamp(t: SensorReading, l: Long): Long = t.timestamp
    })

}

File: src/main/java/util/WatermarkStrategies.java
package util;

import org.apache.flink.api.common.eventtime.WatermarkStrategy;

import java.time.Duration;

public class WatermarkStrategies {

  public WatermarkStrategy<SensorReading> sensorReadingStrategy;

  public WatermarkStrategies() {
    this.sensorReadingStrategy = WatermarkStrategy
        .<SensorReading>forBoundedOutOfOrderness(Duration.ofSeconds(5L))
        .withTimestampAssigner((sr, ts) -> sr.getTimestamp());
  }

}
